### PR TITLE
Script type checks for change-outputs

### PR DESCRIPTION
**Pull request by matejcik**
Issue originally reported by @jhoenicke: Trezors do not do any script type checks on change-outputs, meaning that it is possible to invisibly send change to the right path but wrong script type. This would make the change address invisible in basically every existing wallet software.
This does not cause coin loss because it is relatively easy to recover such "hidden" funds.

### CODE REVIEW

**Comment by matejcik** [comment]
`test_apps.bitcoin.approver.py` is failing because the outputs are all PAYTOWITNESS but expected to be PAYTOTAPROOT. is that a bug in the test? @andrewkozlik

File: legacy/firmware/signing.c [hunk 1]
```diff
@@ -1487,6 +1553,14 @@ static bool signing_validate_input(const TxInputType *txinput) {
     return false;
   }
 
+  if (!coin->decred && txinput->has_decred_tree) {
```
Comment (matejcik) [inline legacy/firmware/signing.c:1556]: should this be under `#if !BITCOIN_ONLY`, just for neatness' sake?

File: legacy/firmware/signing.c [hunk 2]
```diff
@@ -1530,6 +1604,25 @@ static bool signing_validate_input(const TxInputType *txinput) {
...
 
+static bool signing_validate_prev_input(const TxInputType *txinput) {
+  if (txinput->prev_hash.size != 32) {
+    fsm_sendFailure(FailureType_Failure_DataError,
+                    _("Provided prev_hash is invalid."));
+    signing_abort();
+    return false;
+  }
+
+  if (!coin->decred && txinput->has_decred_tree) {
```
Comment (matejcik) [inline legacy/firmware/signing.c:1615]: dtto

**Review by matejcik — COMMENTED** [review]
LGTM overall.

my overarching thought is that sending to a mismatched script type really should be a safety-checks thing, but we don't need to go into that in this PR.

File: legacy/firmware/signing.c (hunk 1 above)
Comment (andrewkozlik) [reply legacy/firmware/signing.c:1556]: Could be, but what should BTC-only firmware do then in `decred_tree` is specified? Should it ignore or return error?

Comment (matejcik) [reply legacy/firmware/signing.c:1556]: ignore bc from the btc-only point of view the field doesn't exist and the protobuf entry is just the kind of noise protobuf is explicitly allowed to have

**Comment by andrewkozlik** [comment]
> sending to a mismatched script type really should be a safety-checks thing

I don't understand.

File: legacy/firmware/signing.c (hunk 1 above)
Comment (andrewkozlik) [reply legacy/firmware/signing.c:1556]: `core` also doesn't take BTC-only into account in this check.

**Comment by andrewkozlik** [comment]
On a different note, once https://github.com/trezor/trezor-firmware/issues/2152 is complete I'd start thinking about changing the logic that a transfer from a multisig account to a singlesig account on the same path is treated as a change-output.

**Review by andrewkozlik — APPROVED** [review]
utACK

File: legacy/firmware/signing.c (hunk 1 above)
Comment (matejcik) [reply legacy/firmware/signing.c:1556]: and it probably also should. but ok, let's keep it the same on both sides

**Comment by matejcik** [comment]
> sending to a mismatched script type really should be a safety-checks thing

like, only allowed when safety checks are lowered.

While there _are_ usecases for using a derivation path with a mismatched script type, I strongly suspect that the vast majority of cases are mistakes that are, for the layperson, difficult to recognize and recover from.
(My model of why this would happen is a misconfigured Electrum or some wallet of that sort. Then once you lose your PC, it will look like the coins are lost, even though recovering them is as simple as redoing the original broken configuration. This is a usecase for descriptors, but nobody is actually backing up their descriptors.)

This is the kind of footgun we should be removing now that we realize we have it lying around.
As a prototype of "good" firmware behavior, a transaction like that should be rejected, ideally before the user even sees anything (which is not currently possible ofc), unless the user has opted into doing weird things.

**Pull request merged**